Spline Utilities

lipohlmann opened the pull request:
closes #30928

## Reason
For use with recently-submitted BSpline utility, which relies on the input of control points to be interpolated. Builds on the project of creating a pipe junction component.

## Design
Utility will auto-create interior control points to be used with BSpline.C. Control points are selected so that the derivatives at supplied points to be connected are honored, which allows the connection to be as smooth as possible. Additional parameter of `sharpness` was included to give the user control over the shape of the curve. Users can also vary the control point density.

## Impact
Supplies general utility for B-Splines (or other spline types) which can have a wide application base.

Commit:
Add control points utility

Created additional utility and tests to create ideal control points for the submitted BSpline utility.
Control points are create such that the supplied derivatives at each point are honored.
User will be able to control the placement of the points and the densitity of these points.

closes #30928

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3005836/), step Docs: sync website on 20565d7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30929/site)

This comment will be updated on new commits.

Commit:
Update BSpline design to provide all parameters in constructor
and use utils to figure out control points

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3005851/), step Generate coverage on 20565d7 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5c8eb0</th>
      <th colspan=4>#30929 20565d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.50%</td>
      <td>85.51%</td>
      <td>+0.01%</td>
      <td>95.54%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>114952</td>
      <td>115060</td>
      <td>+108</td>
      <td>107</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19501</td>
      <td>19505</td>
      <td>+4</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30929/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30929/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30929/coverage/xfem)

</details>



This comment will be updated on new commits.

Commit:
Fix closestPoints method to ensure derivatives are honored correctly

roystgnr commented on framework/include/utils/BSpline.h:
```diff
@@ -0,0 +1,70 @@
+
+#pragma once
+
+#include <vector>
+#include "libmesh/point.h"
```
Let's put headers (here and elsewhere) into the "canonical" order?  In Foo.C, you first include Foo.h; after that (or in Foo.h) you include other MOOSE headers first, then dependency headers (like libmesh/whatever.h) second, then standard headers (like vector) third.

That makes it easier to catch bugs where a header doesn't include all it's proper dependencies but "gets away with it" because those dependencies happen to get included before it in its current use case(s).

roystgnr reviewed (changes requested):
I have a few nitpicks, but the one thing I think is most worth waiting on is the mesh generator using these utilities; it's possible that you'll want to make API changes here before that's completely finished, and if so then better to make them in this PR than in a new one.

If there's reason to merge earlier though, I wouldn't argue.

GiudGiud committed:
Address review:
- use auto and references in utils code
- fix comment character choice
- add more docstring
- use norm_sq where possible to save on sqrt operations

GiudGiud committed:
Address review on unit test:
- use auto when possible
- follow format guide

GiudGiud commented on framework/src/utils/SplineUtils.C:
```diff
@@ -0,0 +1,201 @@
+  std::vector<Point> control_points;
+
+  /// initialize parameter
+  double t;
```
removed

GiudGiud commented on framework/src/utils/SplineUtils.C:
```diff
@@ -0,0 +1,201 @@
+  /// loop over the number of control points to generate the contorl points
+  for (const auto i : make_range(num_cps))
+  {
+    t = (double)i / (double)(num_cps - 1);
```
added

GiudGiud reviewed (approved):
Review should be addressed.
Rebased on top of the first PR

@roystgnr if you want to take a look

GiudGiud committed:
Add a warning for circular regions

GiudGiud committed:
Add an assert to spline utils
Update warning on cps for circular

GiudGiud committed:
Update BSpline unit test to new class design

GiudGiud committed:
Update SplineUtils unit test
- notably allow the warning in circular spline test

GiudGiud committed:
Unity build fix
Unit test throw fixc

roystgnr reviewed (changes requested):
I think we've *almost* got this.

GiudGiud committed:
Address review:
- use square libmesh tolerance for comparing to square norms

Add minor explanation

roystgnr reviewed (approved):
CI failures were unrelated.  I'm kicking the failed recipes because I like to see green, but if they're again thwarted by network/system failures then I'd say let's just merge anyway.
